src/runtime/StateSerialization/MaybeMemberInfo.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.IO;
+
+namespace Python.Runtime
+{
+     [Serializable]
+    internal struct MaybeMemberInfo<T> : ISerializable where T: MemberInfo
+    {
+        public static implicit operator MaybeMemberInfo<T> (T ob) => new MaybeMemberInfo<T>(ob);
+
+        string name;
+        MemberInfo info;
+
+        [NonSerialized]
+        Exception deserializationException;
+
+        public string DeletedMessage 
+        {
+            get
+            {
+                return $"The .NET {typeof(T)} {name} no longer exists. Cause: " + deserializationException?.Message ;
+            }
+        }
+
+        public T Value
+        {
+            get
+            {
+                if (info == null)
+                {
+                    throw new SerializationException(DeletedMessage, innerException: deserializationException);
+                }
+                return (T)info;
+            }
+        }
+
+        public string Name {get{return name;}}
+        public bool Valid => info != null;
+
+        public override string ToString()
+        {
+            return (info != null ? info.ToString() : $"missing type: {name}");
+        }
+
+        public MaybeMemberInfo(T fi)
+        {
+            info = fi;
+            name = info?.ToString();
+            deserializationException = null;
+        }
+
+        internal MaybeMemberInfo(SerializationInfo serializationInfo, StreamingContext context)
+        {
+            // Assumption: name is always stored in "s"
+            name = serializationInfo.GetString("s");
+            info = null;
+            deserializationException = null;
+            try
+            {
+                var tp = Type.GetType(serializationInfo.GetString("t"));
+                if (tp != null)
+                {
+                    var field_name = serializationInfo.GetString("f");
+                    MemberInfo mi = tp.GetField(field_name, ClassManager.BindingFlags);
+                    if (mi != null && ShouldBindMember(mi))
+                    {
+                        info = mi;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                deserializationException = e;
+            }
+        }
+
+        // This is complicated because we bind fields 
+        // based on the visibility of the field, properties 
+        // based on it's setter/getter (which is a method 
+        //  info) visibility and events based on their
+        // AddMethod visibility.
+        static bool ShouldBindMember(MemberInfo mi)
```

Comment:
I'm not totally satisfied with this solution, but it's the best I could find. If anyone has a better idea..

Reply:
You could also leave this abstract and have `class MaybePropertyInfo : MaybeMemberInfo<PropertyInfo>` and so on, enforcing that if the type of member changed then it doesn't count as a match.

Reply:
But maybe that's not critical.